sdk/ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala

```diff
@@ -441,23 +484,40 @@ object PackageService {
   def buildTemplateIdMap[CtId[T] <: ContractTypeId.Definite[T] with ContractTypeId.Ops[CtId, T]](
       idName: PackageNameMap,
       ids: Set[RequiredPkg[CtId]],
-  )(implicit
-      lc: LoggingContextOf[InstanceUUID]
   ): ContractTypeIdMap[CtId] = {
+    import com.daml.nonempty.NonEmptyReturningOps._
     val all = ids.view.map(k => (k, k)).toMap
     val unique = filterUniqueTemplateIs(ids)
-    val nameIds = ids
-      .flatMap(id => idName.get(id.packageId).flatMap(_.toOption).map(n => (n, id.packageId)))
-      .groupBy(_._1)
-      .flatMap {
-        case (_, p) if (p.size == 1) => Some(p.head)
-        case (n, p) => {
-          logger.info(s"Ignoring package name '$n' as we have ${p.size} package ids for it")
-          None
+
+    val idPkgNamePkgVer: Set[(RequiredPkg[CtId], Ref.PackageName, Ref.PackageVersion)] =
+      ids
+        .flatMap { id =>
+          idName
+            .get(id.packageId)
+            .flatMap { case (kpn, v) => kpn.toOption.map(nm => (id, nm, v)) }
+        }
+
+    val nameIds = idPkgNamePkgVer
+      .groupBy1(_._2) // group by package name
+      .map {
+        case (name, idNameVers) => {
+          // Sort the package ids by version, descending
+          val orderedPkgIds: NonEmpty[Seq[String]] = idNameVers
+            .map { case (id, _, ver) => (id.packageId, ver) }
+            .toSeq
+            .sorted(Ordering.by((pkgIdVer: (String, Ref.PackageVersion)) => pkgIdVer._2).reverse)
+            .map(_._1)
+          (name, orderedPkgIds)
         }
       }
       .toMap
-    ContractTypeIdMap(all, unique, nameIds)
+
+    val allByPkgName = idPkgNamePkgVer.map { case (id, name, _) =>
+      val idWithPkgName = id.copy(packageId = s"#${name.toString}")
+      (idWithPkgName, idWithPkgName)
+    }.toMap
+
+    ContractTypeIdMap(all ++ allByPkgName, unique, nameIds, idName)
```

Comment:
As a map (`all + allByPkgName`) this seems to be serving little purpose as everything is mapped back onto itself => could be a Set ?

Reply:
(Looks like it is necessary to get the types to work 🤔 )